File: ai-productivity-app/backend/app/schemas/knowledge.py
"""
Knowledge base API schemas for search and context building.
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field


class KnowledgeSearchRequest(BaseModel):
    """Knowledge base search request."""
    query: str = Field(..., min_length=1)
    project_ids: Optional[List[int]] = None  # Support multiple projects
    filters: Optional[Dict[str, Any]] = None
    limit: Optional[int] = Field(10, ge=1, le=100)
    similarity_threshold: Optional[float] = Field(0.5, ge=0, le=1)
    include_metadata: Optional[bool] = True


class KnowledgeEntry(BaseModel):
    """Knowledge base entry result."""
    id: str
    content: str
    title: Optional[str] = None
    source: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    similarity_score: Optional[float] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ContextBuildRequest(BaseModel):
    """Context building request."""
    knowledge_entries: List[str]  # List of knowledge entry IDs
    query: str
    project_id: str
    max_context_length: Optional[int] = Field(4000, ge=100, le=10000)
    prioritize_recent: Optional[bool] = True


class ContextResult(BaseModel):
    """Built context result."""
    context: str
    sources: List[KnowledgeEntry]
    context_length: int
    relevance_score: float
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class KnowledgeStats(BaseModel):
    """Knowledge base statistics."""
    total_entries: int
    categories: Dict[str, int]
    recent_additions: int
    search_volume: int
    hit_rate: float
    popular_queries: List[Dict[str, Any]]
    last_updated: datetime


class KnowledgeSearchResponse(BaseModel):
    """Knowledge search response."""
    results: List[KnowledgeEntry]
    total_count: int
    query_time: float
    has_more: bool
    suggestions: Optional[List[str]] = None


class QueryAnalysisRequest(BaseModel):
    """Query analysis request."""
    query: str = Field(..., min_length=1)
    project_id: str


class QueryAnalysisResponse(BaseModel):
    """Query analysis response."""
    intent: str = Field(..., description="Detected intent (search, implement, debug, explain, etc.)")
    task_type: str = Field(..., description="Type of task (code_review, documentation, troubleshooting, etc.)")
    complexity: str = Field(..., description="Complexity level (simple, moderate, complex)")
    keywords: List[str] = Field(..., description="Extracted keywords")
    confidence: float = Field(..., ge=0, le=1, description="Analysis confidence score")
    suggested_filters: Optional[Dict[str, Any]] = None


class KnowledgeRetrievalRequest(BaseModel):
    """Knowledge retrieval request."""
    analysis: QueryAnalysisResponse
    project_id: str
    max_docs: Optional[int] = Field(5, ge=1, le=20)
    min_confidence: Optional[float] = Field(0.2, ge=0, le=1)
    auto_context: Optional[bool] = True


class ContextInjectionRequest(BaseModel):
    """Context injection request."""
    query: str = Field(..., min_length=1)
    knowledge: List[KnowledgeEntry]
    citation_style: Optional[str] = Field("inline", pattern="^(inline|footnote)$")
    max_context_length: Optional[int] = Field(2000, ge=100, le=10000)


class ContextInjectionResponse(BaseModel):
    """Context injection response."""
    contextualized_query: str
    context_length: int
    citations_added: int


class CitationRequest(BaseModel):
    """Citation addition request."""
    response: str
    knowledge: List[KnowledgeEntry]
    citation_style: Optional[str] = Field("inline", pattern="^(inline|footnote)$")


class CitationResponse(BaseModel):
    """Citation addition response."""
    response_with_citations: str
    citations: Dict[str, Any]
    citation_count: int


class KnowledgeResponse(BaseModel):
    """Standard knowledge API response."""
    success: bool
    message: Optional[str] = None
    data: Optional[Any] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)

File: ai-productivity-app/frontend/src/api/knowledge.js
/* eslint-disable */
// api/knowledge.js
// Knowledge base and context management API

import client from './client';

export const knowledgeAPI = {
    /**
     * Analyze user query for intent, task type, and keywords
     * @param {string} query - User's input query
     * @param {string} projectId - Project context
     * @returns {Promise<object>} Analysis result with intent, taskType, complexity, keywords
     */
    analyzeQuery: async (query, projectId) => {
        const response = await client.post('/api/knowledge/analyze-query', {
            query,
            project_id: projectId
        });
        return response.data;
    },

    /**
     * Retrieve relevant knowledge from the knowledge base
     * @param {object} analysis - Query analysis result
     * @param {string} projectId - Project context
     * @param {object} settings - Knowledge retrieval settings
     * @returns {Promise<Array>} Array of knowledge documents with confidence scores
     */
    retrieveKnowledge: async (analysis, projectId, settings) => {
        const response = await client.post('/api/knowledge/retrieve', {
            analysis,
            project_id: projectId,
            max_docs: settings.maxContextDocs,
            min_confidence: settings.minConfidence,
            auto_context: settings.autoContext
        });
        return response.data;
    },

    /**
     * Inject context into query for better model understanding
     * @param {string} query - Original query
     * @param {Array} knowledge - Retrieved knowledge documents
     * @param {object} settings - Context injection settings
     * @returns {Promise<string>} Contextualized query
     */
    injectContext: async (query, knowledge, settings) => {
        const response = await client.post('/api/knowledge/inject-context', {
            query,
            knowledge,
            citation_style: settings.citationStyle,
            max_context_length: settings.maxContextDocs * 500 // rough estimate
        });
        return response.data.contextualized_query;
    },

    /**
     * Add citations to the model response
     * @param {object} response - Model response
     * @param {Array} knowledge - Knowledge sources used
     * @param {string} citationStyle - 'inline' or 'footnote'
     * @returns {Promise<object>} Response with citations added
     */
    addCitations: async (response, knowledge, citationStyle) => {
        const apiResponse = await client.post('/api/knowledge/add-citations', {
            response,
            knowledge,
            citation_style: citationStyle
        });
        return apiResponse.data;
    },

    /**
     * Get knowledge base statistics for a project
     * @param {string} projectId - Project ID
     * @returns {Promise<object>} Knowledge base stats
     */
    getKnowledgeStats: async (projectId) => {
        const response = await client.get(`/api/knowledge/stats/${projectId}`);
        return response.data;
    },

    /**
     * Search knowledge base with semantic similarity
     * @param {string} projectId - Project ID
     * @param {string} query - Search query
     * @param {object} options - Search options
     * @returns {Promise<Array>} Search results
     */
    semanticSearch: async (projectId, query, options = {}) => {
        const response = await client.post(`/api/knowledge/search/${projectId}`, {
            query,
            ...options
        });
        return response.data;
    },

    /**
     * Get knowledge base summary for a project
     * @param {string} projectId - Project ID
     * @returns {Promise<object>} Knowledge base summary
     */
    getSummary: async (projectId) => {
        const response = await client.get(`/api/knowledge/summary/${projectId}`);
        return response.data;
    }
};

export default knowledgeAPI;
